### PR TITLE
Cleanup internals

### PR DESCRIPTION
**Pull request by martintmk**
### Details on the issue fix or feature implementation

Cleaning and unifying internal implementation following the recent changes to pipeline composition.

- Introducing an internal `PipelineComponent` that represents an executable unit of a pipeline. It can be created from:
  - ResilienceStrategy (generic/non-generic)
  - ResiliencePipeline (generic/non-generic)
- ResiliencePipeline (generic/non-generic) is now sealed and with internal constructor
- Tidying and unifying some internals
- Dropped `NullResilienceStrategy.Instance` in favor of `ResilienceStrategy.Null` (similar to `Stream.Null`)

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Core/Utils/PipelineComponent.Composite.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        {
+            if (components.Count == 0)
+            {
+                throw new InvalidOperationException("The composite resilience pipeline must contain at least one resilience strategy.");
+            }
+
+            if (components.Distinct().Count() != components.Count)
+            {
+                throw new InvalidOperationException("The composite resilience pipeline must contain unique resilience strategies.");
+            }
```
Comment (martincostello) [inline src/Polly.Core/Utils/PipelineComponent.Composite.cs:44]: Should these exception messages be reworded to talk about components?

File: src/Polly.Core/Utils/PipelineComponent.Composite.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+            if (components.Count == 1)
+            {
+                return new CompositeComponent(components[0], components, telemetry, timeProvider);
+            }
+
+            // convert all strategies to delegating ones (except the last one as it's not required)
+            var delegatingStrategies = components
```
Comment (martincostello) [inline src/Polly.Core/Utils/PipelineComponent.Composite.cs:52]: Refer to components here too?

File: src/Polly.Core/Utils/PipelineComponent.Composite.cs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            _telemetry.Report(
+                new ResilienceEvent(ResilienceEventSeverity.Information, TelemetryUtil.PipelineExecuted),
+                new OutcomeArguments<TResult, PipelineExecutedArguments>(context, outcome, new PipelineExecutedArguments(_timeProvider.GetElapsedTime(timeStamp))));
+
+            return outcome;
+        }
+    }
+
+    /// <summary>
+    /// A resilience strategy that delegates the execution to the next strategy in the chain.
```
Comment (martincostello) [inline src/Polly.Core/Utils/PipelineComponent.Composite.cs:103]: Components

File: src/Polly.Core/Utils/PipelineComponent.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+using System.Threading.Tasks;
+using Polly.Telemetry;
+
+namespace Polly.Utils;
+
+/// <summary>
+/// Represents a single component of a resilience pipeline.
+/// </summary>
+/// <remarks>
+/// The component of the pipeline can be either strategy, generic strategy or a whole pipeline.
```
Comment (martincostello) [inline src/Polly.Core/Utils/PipelineComponent.cs:12]:
```suggestion
/// The component of the pipeline can be either a strategy, a generic strategy or a whole pipeline.
```

File: src/Polly.Core/Utils/TaskHelper.cs [hunk 5]
```diff
@@ -1,7 +1,40 @@
...
 
+#pragma warning disable S5034 // "ValueTask" should be consumed correctly
+
 internal static class TaskHelper
 {
+    public static void GetResult(this ValueTask<VoidResult> task)
+    {
+        Debug.Assert(
+            task.IsCompleted,
+            "The value task should be already completed at this point. If not, it's an indication that the strategy does not respect the ResilienceContext.IsSynchronous value.");
```
Comment (martincostello) [inline src/Polly.Core/Utils/TaskHelper.cs:11]: Does this message need updating?

File: src/Polly.Core/Utils/TaskHelper.cs [hunk 6]
```diff
@@ -1,7 +1,40 @@
...
+        }
+
+        task.Preserve().GetAwaiter().GetResult();
+    }
+
+    public static TResult GetResult<TResult>(this ValueTask<TResult> task)
+    {
+        Debug.Assert(
+            task.IsCompleted,
+            "The value task should be already completed at this point. If not, it's an indication that the strategy does not respect the ResilienceContext.IsSynchronous value.");
```
Comment (martincostello) [inline src/Polly.Core/Utils/TaskHelper.cs:27]: +1

File: src/Polly.Core/Utils/PipelineComponent.Composite.cs (hunk 1 above)
Comment (martintmk) [reply src/Polly.Core/Utils/PipelineComponent.Composite.cs:44]: Not quite sure here. Component is only internal detail while strategy is in the public API. I think both approaches are fine. What's your preference?

File: src/Polly.Core/Utils/TaskHelper.cs (hunk 5 above)
Comment (martintmk) [reply src/Polly.Core/Utils/TaskHelper.cs:11]: I think here it's ok. Component does not care about sync vs async. It's only `ResilienceStrategy` that won't respect this flag.

File: src/Polly.Core/Utils/PipelineComponent.Composite.cs (hunk 1 above)
Comment (martincostello) [reply src/Polly.Core/Utils/PipelineComponent.Composite.cs:44]: Then maybe this is the wrong place to validate this? Feels weird to me for these exception messages to make assumptions about a higher level of the code?

Comment (martintmk) [reply src/Polly.Core/Utils/PipelineComponent.Composite.cs:44]: True, let me move this to pipeline builder.

**Review by martincostello — APPROVED** [review]

**Pull request merged**